### PR TITLE
feature: 044 Creating http clients using axios instances

### PR DESCRIPTION
**Pull request by dtngoc99**
## Description 👋

- Creating HTTP Clients using axios instances

- Create an interceptor to set Auth token into the headers of each request.

- Handling network errors with error interceptor

### CODE REVIEW

File: src/http/ENDPOINTS.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export const AUTHENTICATE_ENDPOINT = "/auth/authenticate";
+export const REGISTER_ENDPOINT = "/auth/register";
+export const COMPANIES_ENDPOINT = "/companies";
```
Comment (luantc13) [inline src/http/ENDPOINTS.js]: Pls follow the message on Slack.

File: src/http/httpClient.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  authInterceptor,
+  errorInterceptor,
+} from "./interceptor";
+
+export const httpClient = axios.create({
+  baseURL: `http://intern-2021.southeastasia.cloudapp.azure.com:4000/api/v1`,
+  timeout: 1000,
+  headers: {
+    Accept: "application/json",
+    "Content-Type": "application/json",
```
Comment (luantc13) [inline src/http/httpClient.js:13]: We should have the same format for all, `"Accept" and "Content-Type"` or `Accept and Content-Type`, just one.

File: src/http/interceptor.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import { REGISTER_ENDPOINT, AUTHENTICATE_ENDPOINT } from "@/http/ENDPOINTS";
+
+const responseInterceptor = (response) => {
+  if (response.status >= 200 && response.status < 300) return response;
+};
```
Comment (luantc13) [inline src/http/interceptor.js:5]: The function content should have `return` for all cases.

File: src/http/interceptor.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+const errorInterceptor = (error) => {
+  if (error.status === 401) {
+    this.notify.warn("Please login again", "Session Expired");
+    localStorage.removeItem("token");
+    this.$store.dispatch("signout/signoutAccount");
+    return Promise.reject(error.status);
+  }
+};
```
Comment (luantc13) [inline src/http/interceptor.js:13]: The same here.

File: src/signin/views/Signin.vue [hunk 5]
```diff
@@ -22,28 +22,18 @@ export default {
...
         })
         .then(() => {
           this.$router.push({ path: "/companies" });
-          NProgress.done();
         })
-        .catch((err) => {
-          this.error = err.response.data.error;
+        .catch(() => {
           this.errorMessage = "Account was wrong!";
           NProgress.done();
```
Comment (luantc13) [inline src/signin/views/Signin.vue:38]: This line is also removed, same as `NProgress.start();` and `NProgress.done();` above.

**Review by luantc13 — CHANGES_REQUESTED** [review]
Great work!
Having some comments, happy to merge when they are resolved.

File: src/signin/views/Signin.vue (hunk 5 above)
Comment (dtngoc99) [reply src/signin/views/Signin.vue:38]: my bad

File: src/http/auth.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import { httpClient } from "@/http/httpClient";
+import { REGISTER_ENDPOINT, AUTHENTICATE_ENDPOINT } from "@/http/ENDPOINTS";
+
+const signin = ({ username, password }) =>
+  httpClient.post(AUTHENTICATE_ENDPOINT, { username, password });
+
+const register = (userInfor) => httpClient.post(REGISTER_ENDPOINT, userInfor);
+export { signin, register };
```
Comment (truonghungit) [inline src/http/auth.js:8]:
```suggestion

export {
  signin,
  register,
};

```
that helps us avoid conflict.

File: src/http/httpClient.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import axios from "axios";
+import {
+  responseInterceptor,
+  authInterceptor,
+  errorInterceptor,
+} from "./interceptor";
+
+export const httpClient = axios.create({
+  baseURL: `http://intern-2021.southeastasia.cloudapp.azure.com:4000/api/v1`,
+  timeout: 1000,
```
Comment (truonghungit) [inline src/http/httpClient.js:10]: May we remove the timeout?

File: src/http/interceptor.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+const authInterceptor = (config) => {
+  if (
+    config.url !== AUTHENTICATE_ENDPOINT ||
+    config.url !== REGISTER_ENDPOINT
+  ) {
+    config.headers.common["Authorization"] = `Bearer ${getAuthToken()}`;
+  }
+  return config;
+};
+export { authInterceptor, errorInterceptor, responseInterceptor };
```
Comment (truonghungit) [inline src/http/interceptor.js:26]:
```suggestion

export { 
  authInterceptor,
  errorInterceptor,
  responseInterceptor
};
```

File: src/http/interceptor.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import { REGISTER_ENDPOINT, AUTHENTICATE_ENDPOINT } from "@/http/endpoints";
+
+const responseInterceptor = (response) => response;
+const errorInterceptor = (error) => {
+  if (error.status === 401) {
```
Comment (truonghungit) [inline src/http/interceptor.js]: I noticed you have an HTTP status code, please use it here
```suggestion
  if (error.status === HTTP_STATUS_CODE.UNAUTHORIZED) {
```

**Review by luantc13 — APPROVED** [review]
Great work!

**Review by truonghungit — CHANGES_REQUESTED** [review]
Good Ngoc.

Please add a change description for the pull request? 
>Example:
> - Creating an HTTP client using Axios
> - Add HTTP interceptor to handle error requests and add an authorization token to the request header
> - Bala bala ...

If we don't have any demo videos, screenshot so delete the #Demo 📺 section

File: src/http/interceptor.js [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import { STATUS } from "@/http/status-code";
+import { REGISTER_ENDPOINT, AUTHENTICATE_ENDPOINT } from "@/http/endpoints";
+
+const responseInterceptor = (response) => response;
+const errorInterceptor = (error) => {
+  if (error.status === STATUS.UNAUTHORIZED) {
+    this.notify.warn("Please login again", "Session Expired");
```
Comment (truonghungit) [inline src/http/interceptor.js:7]: `this` undefined at runtime.

**Review by truonghungit — CHANGES_REQUESTED** [review]
just small issue

File: src/http/interceptor.js (hunk 10 above)
Comment (dtngoc99) [reply src/http/interceptor.js:7]: Do you have any suggest?

**Review by truonghungit — APPROVED** [review]

**Pull request merged**